File: scripts/import_workouts.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"log"
	"os"
	"strings"
	"time"
	"database/sql"
	"github.com/google/uuid"
	_ "github.com/jackc/pgx/v5/stdlib"
)

const (
	DB_URL  = "postgres://vaibhav:@localhost:5432/zenx?sslmode=disable"
	USER_ID = "c4742819-e9aa-4577-98a0-94264070d8af"
	CSV_PATH = "Hevy_workouts_log_100_weeks (1).csv"
)

type ExerciseMapping struct {
	Category  string
	Equipment string
}

func main() {
	db, err := sql.Open("pgx", DB_URL)
	if err != nil {
		log.Fatalf("Failed to connect to database: %v", err)
	}
	defer db.Close()

	if err := db.Ping(); err != nil {
		log.Fatalf("Failed to ping database: %v", err)
	}

	fmt.Println("Connected to database.")

	file, err := os.Open(CSV_PATH)
	if err != nil {
		log.Fatalf("Failed to open CSV: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	header, err := reader.Read()
	if err != nil {
		log.Fatalf("Failed to read header: %v", err)
	}

	columns := make(map[string]int)
	for i, name := range header {
		columns[name] = i
	}

	exercises := make(map[string]string) // name -> id
	
	// Pre-load or seed exercises
	fmt.Println("Processing rows...")

	var currentWorkoutID string
	var currentWorkoutKey string // title + start_time
	var currentWorkoutExerciseID string
	var currentExerciseTitle string
	var exerciseOrder int

	count := 0
	for {
		record, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Fatalf("Failed to read record: %v", err)
		}

		title := record[columns["title"]]
		startTimeStr := record[columns["start_time"]]
		endTimeStr := record[columns["end_time"]]
		workoutDesc := record[columns["description"]]
		exerciseTitle := record[columns["exercise_title"]]
		setIndex := record[columns["set_index"]]
		setType := record[columns["set_type"]]
		weightLbs := record[columns["weight_lbs"]]
		reps := record[columns["reps"]]
		distanceMiles := record[columns["distance_miles"]]
		durationSeconds := record[columns["duration_seconds"]]
		muscleGroup := record[columns["muscle_group"]]

		// 1. Handle Workout
		workoutKey := title + "|" + startTimeStr
		if workoutKey != currentWorkoutKey {
			// New Workout
			startTime, _ := time.Parse("2 Jan 2006, 15:04", startTimeStr)
			// endTime, _ := time.Parse("2 Jan 2006, 15:04", endTimeStr)
			
			// Simple duration calculation if needed
			// dur := int(endTime.Sub(startTime).Minutes())

			err = db.QueryRow(`
				INSERT INTO "workout" ("userId", "name", "description", "date", "notes")
				VALUES ($1, $2, $3, $4, $5)
				RETURNING id`,
				USER_ID, title, workoutDesc, startTime, "").Scan(&currentWorkoutID)
			if err != nil {
				log.Fatalf("Failed to insert workout: %v", err)
			}
			currentWorkoutKey = workoutKey
			exerciseOrder = 0
			currentExerciseTitle = ""
		}

		// 2. Handle Exercise Definition
		exerciseID, exists := exercises[exerciseTitle]
		if !exists {
			// Get or Create Exercise
			err = db.QueryRow("SELECT id FROM exercise WHERE name = $1", exerciseTitle).Scan(&exerciseID)
			if err != nil {
				// Create Exercise
				category := mapMuscleToCategory(muscleGroup)
				equipment := mapTitleToEquipment(exerciseTitle)
				
				err = db.QueryRow(`
					INSERT INTO exercise (name, description, category, "muscleGroup", equipment, "isCustom")
					VALUES ($1, $2, $3, $4, $5, $6)
					RETURNING id`,
					exerciseTitle, "", category, muscleGroup, equipment, false).Scan(&exerciseID)
				if err != nil {
					log.Fatalf("Failed to insert exercise %s: %v", exerciseTitle, err)
				}
			}
			exercises[exerciseTitle] = exerciseID
		}

		// 3. Handle Workout Exercise
		if exerciseTitle != currentExerciseTitle {
			exerciseOrder++
			err = db.QueryRow(`
				INSERT INTO "workout_exercise" ("workoutId", "exerciseId", "order")
				VALUES ($1, $2, $3)
				RETURNING id`,
				currentWorkoutID, exerciseID, exerciseOrder).Scan(&currentWorkoutExerciseID)
			if err != nil {
				log.Fatalf("Failed to insert workout exercise: %v", err)
			}
			currentExerciseTitle = exerciseTitle
		}

		// 4. Handle Set
		// Note: The database uses 'weight' as double precision.
		// If CSV weight is empty, we use 0 or null.
		var weightVal sql.NullFloat64
		if weightLbs != "" {
			var w float64
			fmt.Sscanf(weightLbs, "%f", &w)
			weightVal = sql.NullFloat64{Float64: w, Valid: true}
		}

		var repsVal sql.NullInt32
		if reps != "" {
			var r int32
			fmt.Sscanf(reps, "%d", &r)
			repsVal = sql.NullInt32{Int32: r, Valid: true}
		}

		var durVal sql.NullInt32
		if durationSeconds != "" {
			var d int32
			fmt.Sscanf(durationSeconds, "%d", &d)
			durVal = sql.NullInt32{Int32: d, Valid: true}
		}

		var distVal sql.NullFloat64
		if distanceMiles != "" {
			var d float64
			fmt.Sscanf(distanceMiles, "%f", &d)
			distVal = sql.NullFloat64{Float64: d, Valid: true}
		}

		var setIdxInt int
		fmt.Sscanf(setIndex, "%d", &setIdxInt)

		_, err = db.Exec(`
			INSERT INTO "set" ("workoutExerciseId", "setNumber", "reps", "weight", "duration", "distance", "completed")
			VALUES ($1, $2, $3, $4, $5, $6, $7)`,
			currentWorkoutExerciseID, setIdxInt+1, repsVal, weightVal, durVal, distVal, true)
		if err != nil {
			log.Fatalf("Failed to insert set: %v", err)
		}

		count++
		if count % 100 == 0 {
			fmt.Printf("Processed %d sets...\n", count)
		}
	}

	fmt.Printf("Finished importing %d sets!\n", count)
}

func mapMuscleToCategory(muscle string) string {
	m := strings.ToLower(muscle)
	switch {
	case strings.Contains(m, "chest"):
		return "CHEST"
	case strings.Contains(m, "back") || strings.Contains(m, "lats"):
		return "BACK"
	case strings.Contains(m, "legs") || strings.Contains(m, "quad") || strings.Contains(m, "hamstring") || strings.Contains(m, "calf") || strings.Contains(m, "glute") || strings.Contains(m, "adductor") || strings.Contains(m, "abductor"):
		return "LEGS"
	case strings.Contains(m, "shoulder"):
		return "SHOULDERS"
	case strings.Contains(m, "tricep") || strings.Contains(m, "bicep") || strings.Contains(m, "arm") || strings.Contains(m, "forearm"):
		return "ARMS"
	case strings.Contains(m, "abdominal") || strings.Contains(m, "abs") || strings.Contains(m, "core") || strings.Contains(m, "oblique"):
		return "CORE"
	default:
		return "CORE" // Default to CORE if unknown, or maybe we should have OTHER? 
		// Actually let's check the enums again. CHEST, BACK, LEGS, SHOULDERS, ARMS, CORE.
	}
}

func mapTitleToEquipment(title string) string {
	t := strings.ToLower(title)
	switch {
	case strings.Contains(t, "barbell"):
		return "BARBELL"
	case strings.Contains(t, "dumbbell"):
		return "DUMBBELL"
	case strings.Contains(t, "machine"):
		return "MACHINE"
	case strings.Contains(t, "cable"):
		return "CABLE"
	case strings.Contains(t, "bodyweight") || strings.Contains(t, "box jump") || strings.Contains(t, "push up") || strings.Contains(t, "chin up") || strings.Contains(t, "dip") || strings.Contains(t, "plank") || strings.Contains(t, "yoga"):
		return "BODYWEIGHT"
	default:
		return "OTHER"
	}
}

File: scripts/import_workouts.js
const { Client } = require('pg');
const fs = require('fs');
const path = require('path');

const DB_HOST = process.env.DB_HOST || "localhost";
const DB_BASE_URL = `postgres://zenx:zenx@${DB_HOST}:5432`;
const USER_ID = "38c76f10-20ba-44c9-be55-734070e4fb3a";
const CSV_PATH = path.join(__dirname, '..', 'Hevy_workouts_log_100_weeks (1).csv');

async function createDatabase() {
  console.log(`Checking if database 'zenx' exists on ${DB_HOST}...`);
  const client = new Client({ connectionString: `${DB_BASE_URL}/postgres` });
  await client.connect();

  try {
    const res = await client.query("SELECT 1 FROM pg_database WHERE datname = 'zenx'");
    if (res.rows.length === 0) {
      console.log("Database 'zenx' does not exist. Creating...");
      await client.query("CREATE DATABASE zenx");
      console.log("Database 'zenx' created.");
    } else {
      console.log("Database 'zenx' already exists.");
    }
  } finally {
    await client.end();
  }
}

async function createTables(client) {
  console.log("Creating tables...");
  await client.query('CREATE EXTENSION IF NOT EXISTS "uuid-ossp";');

  await client.query(`CREATE TABLE IF NOT EXISTS "workout" (
        "id" UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
        "userId" UUID NOT NULL,
        "name" TEXT NOT NULL,
        "description" TEXT,
        "date" TIMESTAMPTZ,
        "notes" TEXT,
        "likes" TEXT[] DEFAULT '{}',
        "created_at" TIMESTAMPTZ DEFAULT NOW()
    );`);

  await client.query(`CREATE TABLE IF NOT EXISTS "exercise" (
        "id" UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
        "name" TEXT NOT NULL,
        "description" TEXT,
        "category" TEXT,
        "muscleGroup" TEXT,
        "equipment" TEXT,
        "isCustom" BOOLEAN DEFAULT FALSE,
        "created_at" TIMESTAMPTZ DEFAULT NOW()
    );`);

  await client.query(`CREATE TABLE IF NOT EXISTS "workout_exercise" (
        "id" UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
        "workoutId" UUID REFERENCES "workout"("id") ON DELETE CASCADE,
        "exerciseId" UUID REFERENCES "exercise"("id"),
        "order" INT,
        "notes" TEXT,
        "created_at" TIMESTAMPTZ DEFAULT NOW()
    );`);

  await client.query(`CREATE TABLE IF NOT EXISTS "set" (
        "id" UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
        "workoutExerciseId" UUID REFERENCES "workout_exercise"("id") ON DELETE CASCADE,
        "setNumber" INT,
        "reps" INT,
        "weight" FLOAT,
        "duration" INT,
        "distance" FLOAT,
        "completed" BOOLEAN DEFAULT TRUE,
        "created_at" TIMESTAMPTZ DEFAULT NOW()
    );`);
  console.log("Tables created.");
}

function parseCSV(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  const lines = content.split('\n');
  const header = lines[0].split(',');
  const result = [];

  for (let i = 1; i < lines.length; i++) {
    if (!lines[i]) continue;

    // Handle cases where commas are inside quotes (like start_time)
    const row = [];
    let current = '';
    let inQuotes = false;
    for (let char of lines[i]) {
      if (char === '"') {
        inQuotes = !inQuotes;
      } else if (char === ',' && !inQuotes) {
        row.push(current);
        current = '';
      } else {
        current += char;
      }
    }
    row.push(current);

    const obj = {};
    header.forEach((h, index) => {
      obj[h.trim()] = row[index] ? row[index].trim() : '';
    });
    result.push(obj);
  }
  return result;
}

function mapMuscleToCategory(muscle) {
  const m = (muscle || '').toLowerCase();
  if (m.includes('chest')) return 'CHEST';
  if (m.includes('back') || m.includes('lats')) return 'BACK';
  if (m.includes('legs') || m.includes('quad') || m.includes('hamstring') || m.includes('calf') || m.includes('glute') || m.includes('adductor') || m.includes('abductor')) return 'LEGS';
  if (m.includes('shoulder')) return 'SHOULDERS';
  if (m.includes('tricep') || m.includes('bicep') || m.includes('arm') || m.includes('forearm')) return 'ARMS';
  if (m.includes('abdominal') || m.includes('abs') || m.includes('core') || m.includes('oblique')) return 'CORE';
  return 'CORE';
}

function mapTitleToEquipment(title) {
  const t = (title || '').toLowerCase();
  if (t.includes('barbell')) return 'BARBELL';
  if (t.includes('dumbbell')) return 'DUMBBELL';
  if (t.includes('machine')) return 'MACHINE';
  if (t.includes('cable')) return 'CABLE';
  if (t.includes('bodyweight') || t.includes('box jump') || t.includes('push up') || t.includes('chin up') || t.includes('dip') || t.includes('plank') || t.includes('yoga')) return 'BODYWEIGHT';
  return 'OTHER';
}

async function run() {
  await createDatabase();

  const client = new Client({
    connectionString: `${DB_BASE_URL}/zenx`,
  });

  await client.connect();
  console.log('Connected to zenx database.');

  await createTables(client);

  // Check if data already exists
  const checkRes = await client.query('SELECT count(*) FROM "workout"');
  if (parseInt(checkRes.rows[0].count) > 0) {
    console.log(`Database already has ${checkRes.rows[0].count} workouts. Skipping import.`);
    await client.query('TRUNCATE "workout", "exercise", "workout_exercise", "set" CASCADE');
    console.log('Truncated existing data for fresh import.');
  }

  const data = parseCSV(CSV_PATH);
  console.log(`Parsed ${data.length} rows from CSV.`);

  const exerciseMap = new Map(); // name -> id

  let currentWorkoutID = null;
  let currentWorkoutKey = null;
  let currentWorkoutExerciseID = null;
  let currentExerciseTitle = null;
  let exerciseOrder = 0;

  for (let i = 0; i < data.length; i++) {
    const row = data[i];

    // 1. Workout
    const workoutKey = row.title + '|' + row.start_time;
    if (workoutKey !== currentWorkoutKey) {
      // New Workout
      const date = new Date(row.start_time);

      const res = await client.query(
        'INSERT INTO "workout" ("userId", "name", "description", "date", "notes", "likes") VALUES ($1, $2, $3, $4, $5, $6) RETURNING id',
        [USER_ID, row.title, row.description || '', date, '', []]
      );
      currentWorkoutID = res.rows[0].id;
      currentWorkoutKey = workoutKey;
      exerciseOrder = 0;
      currentExerciseTitle = null;
      // console.log(`Started workout: ${row.title} (${date.toISOString()})`);
    }

    // 2. Exercise
    let exerciseID = exerciseMap.get(row.exercise_title);
    if (!exerciseID) {
      // Check if exists in DB
      const exRes = await client.query('SELECT id FROM "exercise" WHERE name = $1', [row.exercise_title]);
      if (exRes.rows.length > 0) {
        exerciseID = exRes.rows[0].id;
      } else {
        // Create Exercise
        const category = mapMuscleToCategory(row.muscle_group);
        const equipment = mapTitleToEquipment(row.exercise_title);
        const newExRes = await client.query(
          'INSERT INTO "exercise" ("name", "description", "category", "muscleGroup", "equipment", "isCustom") VALUES ($1, $2, $3, $4, $5, $6) RETURNING id',
          [row.exercise_title, '', category, row.muscle_group, equipment, false]
        );
        exerciseID = newExRes.rows[0].id;
      }
      exerciseMap.set(row.exercise_title, exerciseID);
    }

    // 3. Workout Exercise
    if (row.exercise_title !== currentExerciseTitle) {
      exerciseOrder++;
      const weRes = await client.query(
        'INSERT INTO "workout_exercise" ("workoutId", "exerciseId", "order", "notes") VALUES ($1, $2, $3, $4) RETURNING id',
        [currentWorkoutID, exerciseID, exerciseOrder, '']
      );
      currentWorkoutExerciseID = weRes.rows[0].id;
      currentExerciseTitle = row.exercise_title;
    }

    // 4. Set
    const weightVal = row.weight_lbs ? parseFloat(row.weight_lbs) : null;
    const repsVal = row.reps ? parseInt(row.reps) : null;
    const durVal = row.duration_seconds ? parseInt(row.duration_seconds) : null;
    const distVal = row.distance_miles ? parseFloat(row.distance_miles) : null;
    const setNum = parseInt(row.set_index) + 1;

    await client.query(
      'INSERT INTO "set" ("workoutExerciseId", "setNumber", "reps", "weight", "duration", "distance", "completed") VALUES ($1, $2, $3, $4, $5, $6, $7)',
      [currentWorkoutExerciseID, setNum, repsVal, weightVal, durVal, distVal, true]
    );

    if (i % 500 === 0) {
      console.log(`Processed ${i} rows...`);
    }
  }

  console.log('Import finished successfully!');
  await client.end();
}

run().catch(err => {
  console.error(err);
  process.exit(1);
});
